File: backend/src/routes/documents.ts
import { Router } from "express";
import multer from "multer";
import rateLimit from "express-rate-limit";
import fs from "node:fs";
import path from "node:path";
import { z } from "zod";
import { verifyDocument, DOCUMENT_TYPES, type DocumentType } from "../services/documentVerification";
import { buildDocumentChecklist, type DocumentRecord } from "../services/documentChecklist";
import { getDb, getStudentBySessionId, appendDocumentRecord } from "../services/db";
import { studentRowToProfile, type StudentProfile } from "../services/leadScoring";
import { config } from "../config";

export const documentsRouter = Router();

// Vision calls are more expensive than a text chat turn — keep this tight.
const documentsLimiter = rateLimit({
  windowMs: 60 * 1000,
  max: config.documentRateLimitMax,
  standardHeaders: true,
  legacyHeaders: false,
});

// Read-only, no Claude call — lighter than documentsLimiter, but still
// scoped to its own route (never the shared /api prefix) per convention.
const checklistLimiter = rateLimit({
  windowMs: 60 * 1000,
  max: 30,
  standardHeaders: true,
  legacyHeaders: false,
});

function parseDocumentRecords(documentsJson: string | null): DocumentRecord[] {
  if (!documentsJson) return [];
  return JSON.parse(documentsJson) as DocumentRecord[];
}

const ALLOWED_MIME_TYPES = new Set(["image/jpeg", "image/png", "image/webp", "application/pdf"]);
const MAX_FILE_BYTES = 10 * 1024 * 1024; // 10MB

const upload = multer({
  storage: multer.memoryStorage(),
  limits: { fileSize: MAX_FILE_BYTES },
  fileFilter: (_req, file, cb) => {
    if (!ALLOWED_MIME_TYPES.has(file.mimetype)) {
      cb(new Error(`Unsupported file type: ${file.mimetype}. Allowed: JPEG, PNG, WEBP, PDF.`));
      return;
    }
    cb(null, true);
  },
});

const uploadFieldsSchema = z.object({
  documentType: z.enum(DOCUMENT_TYPES as [DocumentType, ...DocumentType[]]),
  sessionId: z.string().min(1).max(200),
});

// SessionId is client-generated (crypto.randomUUID()) — never trust it
// directly in a filesystem path. Strip to a safe charset before using it as
// a directory name.
function sanitizeForPath(value: string): string {
  return value.replace(/[^a-zA-Z0-9-]/g, "").slice(0, 100) || "unknown-session";
}

function extensionFor(mimeType: string): string {
  switch (mimeType) {
    case "image/jpeg":
      return "jpg";
    case "image/png":
      return "png";
    case "image/webp":
      return "webp";
    case "application/pdf":
      return "pdf";
    default:
      return "bin";
  }
}

documentsRouter.post("/documents/upload", documentsLimiter, upload.single("file"), async (req, res) => {
  if (!req.file) {
    res.status(400).json({ error: "No file uploaded (expected multipart field 'file')" });
    return;
  }

  const parsed = uploadFieldsSchema.safeParse(req.body);
  if (!parsed.success) {
    res.status(400).json({ error: "Invalid request fields", details: parsed.error.flatten() });
    return;
  }
  const { documentType, sessionId } = parsed.data;

  try {
    const student = getStudentBySessionId(getDb(), sessionId);
    const profile: StudentProfile | undefined = student ? studentRowToProfile(student) : undefined;

    const result = await verifyDocument(documentType, req.file.buffer, req.file.mimetype, profile);

    const safeSessionDir = sanitizeForPath(sessionId);
    const uploadDir = path.join(config.uploadsDir, safeSessionDir);
    fs.mkdirSync(uploadDir, { recursive: true });
    const filename = `${documentType}-${Date.now()}.${extensionFor(req.file.mimetype)}`;
    fs.writeFileSync(path.join(uploadDir, filename), req.file.buffer);

    const record = {
      documentType,
      filename,
      status: result.status,
      issues: result.issues,
      extractedSummary: result.extractedSummary,
      uploadedAt: new Date().toISOString(),
    };
    const updated = appendDocumentRecord(getDb(), sessionId, record);
    const checklist = buildDocumentChecklist(parseDocumentRecords(updated?.documents ?? null));

    console.log(`[documents] session=${sessionId} type=${documentType} status=${result.status}`);

    res.json({ documentType, ...result, sessionId, checklist });
  } catch (err) {
    console.error("[documents/upload] request failed:", err);
    res.status(500).json({ error: "Failed to verify document. Please try again." });
  }
});

documentsRouter.get("/documents/checklist/:sessionId", checklistLimiter, (req, res) => {
  const sessionId = req.params.sessionId;
  const student = getStudentBySessionId(getDb(), sessionId);
  if (!student) {
    // Not an error — a student who hasn't submitted their profile yet simply
    // has all documents outstanding.
    res.json({ sessionId, checklist: buildDocumentChecklist([]) });
    return;
  }
  const checklist = buildDocumentChecklist(parseDocumentRecords(student.documents));
  res.json({ sessionId, checklist });
});

File: backend-py/app/routers/documents.py
import json
import logging
import re
import time
from pathlib import Path

from fastapi import APIRouter, File, Form, HTTPException, Request, UploadFile

from app.limiter import limiter
from app.config import config
from app.db import append_document_record, ensure_student_stub, get_db, get_student_by_session_id
from app.services.document_checklist import build_document_checklist
from app.services.document_verification import DOCUMENT_TYPES, verify_document
from app.services.lead_scoring import student_row_to_profile

router = APIRouter()
logger = logging.getLogger("aiec.documents")

_ALLOWED_MIME_TYPES = {"image/jpeg", "image/png", "image/webp", "application/pdf"}
_MAX_FILE_BYTES = 10 * 1024 * 1024  # 10MB


def _parse_document_records(documents_json: str | None) -> list[dict]:
    return json.loads(documents_json) if documents_json else []


def _sanitize_for_path(value: str) -> str:
    return re.sub(r"[^a-zA-Z0-9-]", "", value)[:100] or "unknown-session"


def _extension_for(mime_type: str) -> str:
    return {
        "image/jpeg": "jpg",
        "image/png": "png",
        "image/webp": "webp",
        "application/pdf": "pdf",
    }.get(mime_type, "bin")


@router.post("/documents/upload")
@limiter.limit(f"{config.document_rate_limit_max}/minute")
async def upload_document(
    request: Request,
    file: UploadFile = File(...),
    documentType: str = Form(...),
    sessionId: str = Form(..., min_length=1, max_length=200),
):
    if documentType not in DOCUMENT_TYPES:
        raise HTTPException(status_code=400, detail="Invalid request fields")
    if file.content_type not in _ALLOWED_MIME_TYPES:
        raise HTTPException(
            status_code=400,
            detail=f"Unsupported file type: {file.content_type}. Allowed: JPEG, PNG, WEBP, PDF.",
        )

    contents = await file.read()
    if len(contents) > _MAX_FILE_BYTES:
        raise HTTPException(status_code=400, detail="File too large (max 10MB)")

    try:
        conn = get_db()
        # Guarantees a row exists before the document is attached to it — see
        # ensure_student_stub()'s docstring for why this matters even though
        # the real UI always submits a profile first.
        student = ensure_student_stub(conn, sessionId)
        profile = student_row_to_profile(student)

        result = verify_document(documentType, contents, file.content_type, profile)

        safe_session_dir = _sanitize_for_path(sessionId)
        upload_dir = config.uploads_dir / safe_session_dir
        upload_dir.mkdir(parents=True, exist_ok=True)
        filename = f"{documentType}-{int(time.time() * 1000)}.{_extension_for(file.content_type)}"
        (upload_dir / filename).write_bytes(contents)

        from app.db import now_iso

        record = {
            "documentType": documentType,
            "filename": filename,
            "status": result["status"],
            "issues": result["issues"],
            "extractedSummary": result["extractedSummary"],
            "uploadedAt": now_iso(),
        }
        updated = append_document_record(conn, sessionId, record)
        checklist = build_document_checklist(_parse_document_records(updated["documents"] if updated else None))

        logger.info("[documents] session=%s type=%s status=%s", sessionId, documentType, result["status"])

        return {"documentType": documentType, **result, "sessionId": sessionId, "checklist": checklist}
    except HTTPException:
        raise
    except Exception:
        logger.exception("[documents/upload] request failed")
        raise HTTPException(status_code=500, detail="Failed to verify document. Please try again.")


@router.get("/documents/checklist/{session_id}")
@limiter.limit("30/minute")
def get_checklist(request: Request, session_id: str):
    student = get_student_by_session_id(get_db(), session_id)
    if not student:
        # Not an error — a student who hasn't submitted their profile yet
        # simply has all documents outstanding.
        return {"sessionId": session_id, "checklist": build_document_checklist([])}
    checklist = build_document_checklist(_parse_document_records(student["documents"]))
    return {"sessionId": session_id, "checklist": checklist}
